Add documentation (#87)
precice/systemtests: docs/architecture.md

# General architecture

When running on Travis tests are executed in the several stages that are outlined below. 
When it comes to local tests, steps are identical, but the push to the remote dockerhub repository is absent.

## Building preCICE

Several images of preCICE are built, based on different systems (Ubuntu 16.04/18.04, Arch Linux) and different specifications
of properties of the image ( root/non-root, package, different MPI versions ). This corresponds to the `Dockerfile.$systemname$.${system_spec}`

> Resultant images are then pushed to the [Dockerhub](https://hub.docker.com/u/precice).

Reviewer: ```suggestion

Resulting images are then pushed to the [Dockerhub](https://hub.docker.com/u/precice).
```

## Building solvers

> Is done once for every solver, using `Dockerfile.${solver_name}`. Resultant images are then stored on the [Dockerhub](https://hub.docker.com/u/precice).

Reviewer: ```suggestion
Is done once for every solver, using `Dockerfile.${solver_name}`. 

Resulting images are then stored on the [Dockerhub](https://hub.docker.com/u/precice).
```

## Building adapters

To build an adapter we follow a two-stage build, first importing from the preCICE image and then importing from the solver's image
(if the solver's image is actually relevant for building the adapter).We can import from different preCICE base images by specifying `from` 
argument during building of the image, e.g:
```
docker build --build-arg from=precice/precice-ubuntu1804.home-develop -d Dockerfile.openfoam-adapter -t openfoam-adapter-ubuntu1804.home-develop .
```
Note, that due to the fact that each adapter "adapts" differently, the general procedure is slightly different and is outlined below:

- **SU2 adapter**
  Since the SU2 adapter modifies the SU2 source code, it makes no sense to copy a clean SU2 build, because it will need to be rebuilt. Instead we directly build adapted
  SU2 source.

- **deal.II adapter**

>   Since deal.II is a library and the adapter just needs to link with it, here we actually perform the import of deal.II image and only build

Reviewer: ```suggestion
  Since deal.II is a library and the adapter just needs to link to it. Thus, we perform the import of deal.II image and only build
```
